numpy/lib/function_base.py

```diff
@@ -979,13 +979,18 @@ def gradient(f, *varargs, **kwargs):
         # scalar or 1d array for each axis
         dx = list(varargs)
         for i, distances in enumerate(dx):
-            if np.ndim(distances) == 0:
+            distances = np.asarray(distances)
+            if distances.ndim == 0:
                 continue
-            elif np.ndim(distances) != 1:
+            elif distances.ndim != 1:
                 raise ValueError("distances must be either scalars or 1d")
             if len(distances) != f.shape[axes[i]]:
                 raise ValueError("when 1d, distances must match "
                                  "the length of the corresponding dimension")
+            if np.issubdtype(distances.dtype, np.integer):
+                # Convert numpy integer types to float64 to avoid modular
+                # arithmetic in np.diff(distances).
+                distances = distances.astype(np.float64)
             diffx = np.diff(distances)
```

Comment:
Previously `distances` was an ndarray subclass or array_like on this line, now it isn't. Since it's not part of the bug fix, can we back out the ndim changes above?

Reply:
Would it be sufficient to replace `distances = np.asarray(distances)` with `distances = np.asanyarray(distances)`?

Reply:
Answering my own question... almost the first thing `diff` does is pass its argument through `asanyarray`, so it should fine to do that in `gradient`, too.

Reply:
`asanyarray` is probably safe then - although it still belongs in a separate commit (or even PR) since it's unrelated cleanup.

Reply:
The `dtype` attribute of the result is used to determine if `distances` is an integer array, so it is not unrelated.

Reply:
Thanks, that's what I missed.